first attempt

Shushania opened the pull request.

Shushania committed:
first attempt

Alviner commented on notificator/ADR/adr-001.md:
```diff
@@ -0,0 +1,27 @@
+* Простота в использование
+* Универсальность работы
+* Возможности масштабирования
+* Дополнительная защита, где можно обрабатывать ошибки
```
использование rest-api не дает возможности масштабирования и защиты, кажется это из другого ard просочилось

Shushania replied:
Тут два пункта, прокси и тест апи, этот пункт относится к прокси

Alviner replied:
API Gateway в ard002 указан, поэтому и вопрос возник, в adr001 не нашел про прокси, только в отрицительных последствиях есть упоминание

Alviner commented on notificator/ADR/adr-002.md:
```diff
@@ -0,0 +1,24 @@
+
+#### Контекст:
+Для единой точки входа в сервис и балансировки нагрузки предлагается использовать Nginx.
+Причина: распространенность технологии 
```
Указанная причина не покрывает обоснованность использования этого контейнера.

Можно также брать любое из топ 5, тот-же HAProxy с такой-же причиной. Тут скорее нужно оглядываться на опыт команды, указанных требований, сроки и бюджет

Alviner commented on notificator/ADR/adr-003.md:
```diff
@@ -0,0 +1,26 @@
+#### Контекст:
+Необходимо хранить конфигурации для отправки уведомлений
+
+Нужен быстрый поиск по данным, для этого нужна реляционная система.
```
слово быстро очень опасное, что значит быстро? 1секунда это быстро? а 2? Чем скорость нереляционной бд тут проигрывать?

Alviner commented on notificator/ADR/adr-003.md:
```diff
@@ -0,0 +1,26 @@
+Необходимо хранить конфигурации для отправки уведомлений
+
+Нужен быстрый поиск по данным, для этого нужна реляционная система.
+Сильных ограничений на базу в данном случае нет. Выбираем Постгрес, как опенсорсное и универсальное решение
```
использовать pg действительно быстро, а вот администрировать совсем нет.

Тут также стоит оглядываться на команду, может уже есть как опыт так и построенные процессы.

Если команда работала только в ydb, будет странным решением тащить ее в pg чтобы хранить настройки

Alviner commented on notificator/ADR/adr-004.md:
```diff
@@ -0,0 +1,42 @@
+
+#### Последствия:
+##### Положительные:
+* гарантия сохранности оповещений при сбоях
```
прям гаратний в rabbitmq нет, например сообщение не сразу попадает на диск если указана персистентная очередь, поэтому потерять сообщения вполне себе можно. 

Также нельзя исключать нештатные ситуации, когда данные могут не могут быть восстановлены и часть будет потеряна.

Alviner commented on notificator/ADR/adr-004.md:
```diff
@@ -0,0 +1,42 @@
+* возможность увеличения количества очередей обработки
+
+##### Отрицательные:
+* Дополнительная нагрузка на поддержание
```
еще не забудьте про сопровождение, поддержка это что-то точечное и адресное, а сопровождение идет непрерывно

Alviner commented on notificator/ADR/adr-006.md:
```diff
@@ -0,0 +1,27 @@
+* Трошина Дарья
+
+#### Решения:
+* Использования различных топиков для различных каналов
```
отлично, молодец! действительно в указанном контейнере будет возможность хорошо масштабироваться

Alviner commented on notificator/static/c2.png:
Супер, что используется стандартная нотация. 
Второй уровень в моделях использует абстрацию контейнер.

Вообще весь этот подход похож на карту, вот у нас есть страны (системы) и погран пункты между ними.
При увеличение зума есть регионы (контейнеры) и магистральные дороги между ними.
Еще при увеличении есть города (компоненты) и обычные дороги между между ними.

В предложенном варианте слишком подробная схема это больше на третий уровень уже похоже.

Shushania replied:
Я могу путать, но вроде я так и старалась сделать, по сути микросервисы, где у меня api gateway, потом что то что обработает и преобразует в сообщеньку. Потом очередь и свои обработки на каждый канал . Это если верхоуровнево. Или стоит примерно на остановится на Апи, обработчике в сообщеньку и хэндлере на канал? Тут базы лишние? Я просто хотела показать, что база за пределом сервиса в целом.

Alviner replied:
база как раз нужна, скорее просочились компоненты-handlers на схему контейнеров

Alviner commented on notificator/static/c3_1.png:
вот тут уже интересная агрегация, есть же notificaion service, но его нет на уровне выше

Shushania replied:
В названиях запуталась

Alviner reviewed (changes requested):
Здравствуйте!
Извините за задержку проверки, были технические заминки.
Отличная работа! Мне понравились ard, за счет того, что они не перегружены.

Но есть небольшое замечание, которое нужно доработать:
- не нашел описание специфики данных, отсюда и неясна часть принятых решений
- в некоторых ard не хватает или не достаточно весомые обоснования
- отсутствует описание компонентов системы

Будет только плюсом указать еще краткое описание использованных паттернов и описание их имплементации

Более подробно эти и более мелкие мелкие замечания, а так же места в которых можно улучшить проект, отмечены в пр

Напоминаю, что работа не может быть принята без исправления всех критических ошибок! Желаю удачи!

Shushania committed:
second attempt

- change diagrams
- add explanation in adrs
- split adr rest api and proxy
- correct misundestanding
- describe databases and data inside

Shushania committed:
second attempt

- rest api

Shushania commented:
Постаралась Исправить все, но возможно остались ко мне вопросы, где я чего то не раскрыла

Shushania committed:
second attempt

Alviner commented:
> Постаралась Исправить все, но возможно остались ко мне вопросы, где я чего то не раскрыла

не стесняйся при наличии вопросов можешь в пачке открывать тред в чате общем или приходить в личку)

Alviner commented on notificator/ADR/adr-004.md:
```diff
@@ -0,0 +1,42 @@
+* ActiveMQ
+* HiveMQ
+
+Kafka - нет решений для маршрутизации из коробки, необходимы дополнительные решения
```
в кафке есть топики, поэтому можно писать сразу в нужный)

Shushania replied:
Это кусок из теории Практикума, но видно имелось в виду, что надо создавать каждый топик, настраивать, это дополнительные затраты. Ну или теорию стоит переписать

Alviner commented on notificator/ADR/adr-004.md:
```diff
@@ -0,0 +1,42 @@
+Kafka - нет решений для маршрутизации из коробки, необходимы дополнительные решения
+ActiveMQ - сложен в использование, недостаточно подробная документация
+RabbitMQ - проще в использовании, хорошая документация, меньшая производительность
+HiveMQ - не хранит историю сообщений, поэтому не реализует требование 5
```
rabbitmq тоже не хранит историю, после обработки сообщение отбрасывается

Alviner reviewed (changes requested):
Здравствуйте!
Ознакомился с правками!

Стало лучше, идешь верной дорогой. Молодец!

Давай еще скорректируем некоторые моменты:
- adr это скорее лог принятие решений, основной документ все-таки `notificator/software_architecture_document.md`, который должен самодостаточно содержать требования об ожидаемой нагрузки на основании чего можно будет делать предположения, а также текущее состояние системы, с описанием компонент, их взаимодействия
- еще можно добавить в adr к компонентам системы ссылки на материалы об особенностях, особенно с имплементацией выбранных паттернов
- на основании схемы и описания будут строить систему, поэтому необходимы критерии оценки прототипа

Также стоит проработать такую ситуацию: часто сервисы, которые предоставляют услуги отправки пушей предоставляют отправку сразу нескольких сообщений в одном запросе, но сильно ограничивают количество запросов в единицу времени, поэтому лимиты отличны в зависимости от канала передачи.

Напоминаю, что работа не может быть принята без исправления всех критических ошибок! Немного осталось, удачи!

Shushania commented:
по предложению по лимитам : Подразумевается, что лимитами пушей будет заниматься кусок, который называется push service. И на с2 я как раз смс емейл и пуши разделяю, поэтому и настройки соответсвенно разные. Но я этого не пояснила - моя вина. Достаточно будет ли такого пояснения? или в целом подход не верный?

Shushania committed:
third attempt

Shushania committed:
third attempt

Shushania committed:
third attempt

Alviner reviewed (approved):
Отлично! Твой проект подсистемы для отправки оповещений о событиях выглядит многообещающе.

Ты успешно реализовала функциональность для рассылки уведомлений через различные каналы связи, включая email, SMS и push-уведомления. Возможность настройки для каждого канала добавляет гибкости в использовании системы.

Документ проекта неплохо структурирован, что облегчит его понимание и поддержку. Спасибо за твой вклад в проект, и продолжай в том же духе!
